src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/ReactiveCosmosTemplate.java

```diff
@@ -0,0 +1,497 @@
+import com.microsoft.azure.cosmos.CosmosItem;
+import com.microsoft.azure.cosmos.CosmosItemRequestOptions;
+import com.microsoft.azure.cosmosdb.FeedOptions;
+import com.microsoft.azure.cosmosdb.FeedResponse;
+import com.microsoft.azure.cosmosdb.PartitionKey;
+import com.microsoft.azure.cosmosdb.SqlQuerySpec;
+import com.microsoft.azure.documentdb.Document;
+import com.microsoft.azure.spring.data.cosmosdb.CosmosDbFactory;
+import com.microsoft.azure.spring.data.cosmosdb.core.convert.MappingDocumentDbConverter;
+import com.microsoft.azure.spring.data.cosmosdb.core.generator.CountQueryGenerator;
+import com.microsoft.azure.spring.data.cosmosdb.core.generator.FindQuerySpecGenerator;
+import com.microsoft.azure.spring.data.cosmosdb.core.query.Criteria;
+import com.microsoft.azure.spring.data.cosmosdb.core.query.CriteriaType;
+import com.microsoft.azure.spring.data.cosmosdb.core.query.DocumentQuery;
+import com.microsoft.azure.spring.data.cosmosdb.exception.DocumentDBAccessException;
+import com.microsoft.azure.spring.data.cosmosdb.repository.support.DocumentDbEntityInformation;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.json.JSONObject;
+import org.springframework.beans.BeansException;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.lang.NonNull;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class ReactiveCosmosTemplate implements ReactiveCosmosOperations, ApplicationContextAware {
+    private static final String COUNT_VALUE_KEY = "_aggregate";
+
+    private final MappingDocumentDbConverter mappingDocumentDbConverter;
+    private final String databaseName;
+
+    @Getter(AccessLevel.PRIVATE)
+    private final CosmosClient cosmosClient;
+    private final CosmosDbFactory cosmosDbFactory;
+
+    private List<String> collectionCache;
+
+    /**
+     * Constructor
+     *
+     * @param cosmosDbFactory            the cosmosdbfactory
+     * @param mappingDocumentDbConverter the mappingDocumentDbConverter
+     * @param dbName                     database name
+     */
+    public ReactiveCosmosTemplate(CosmosDbFactory cosmosDbFactory,
+                                  MappingDocumentDbConverter mappingDocumentDbConverter,
+                                  String dbName) {
+        Assert.notNull(cosmosDbFactory, "CosmosDbFactory must not be null!");
+        Assert.notNull(mappingDocumentDbConverter, "MappingDocumentDbConverter must not be null!");
+
+        this.databaseName = dbName;
+        this.mappingDocumentDbConverter = mappingDocumentDbConverter;
+        this.collectionCache = new ArrayList<>();
```

Comment:
When using caches in a reactive API, they should be thread-safe to avoid race conditions and visibility issues.